testplan/report/testing/base.py

```diff
@@ -62,56 +63,80 @@
 from testplan.testing import tagging
 
 
+class RuntimeStatus(object):
+    """
+    Constants for test runtime status - for interactive mode
+    """
+
+    READY = 'ready'
+    RUNNING = 'running'
+    FINISHED = 'finished'
+
+    STATUS_PRECEDENCE = (
+        READY,
+        RUNNING,
+        FINISHED,
+        None
+    )
+
+    @classmethod
+    def precedent(cls, stats, rule=STATUS_PRECEDENCE):
+        """
+        Return the precedent status from a list of statuses, using the
+        ordering of statuses in `rule`.
+
+        :param stats: List of statuses of which we want to get the precedent.
+        :type stats: ``sequence``
+        :param rule: Precedence rules for the given statuses.
+        :type rule: ``sequence``
+        """
+        return min(
+            stats,
+            key=lambda stat: rule.index(stat)
+        )
+
+
 class Status(object):
     """
-    Report status constants and utilities for getting precedent statuses.
+    Constants for test result and utilities for propagating status upward.
     """
     ERROR = 'error'
     FAILED = 'failed'
     INCOMPLETE = 'incomplete'
     PASSED = 'passed'
     SKIPPED = 'skipped'
-    READY = 'ready'
-    RUNNING = 'running'
     XFAIL = 'xfail'
     XPASS = 'xpass'
+    XPASS_STRICT = 'xpass-strict'
+    UNSTABLE = 'unstable'
+    UNKNOWN = 'unknown'
 
+    # We only maintain precedence among these primary status
     STATUS_PRECEDENCE = (
-        RUNNING,
-        READY,
-        ERROR,
-        FAILED,
-        INCOMPLETE,
-        PASSED,
-        XPASS,
-        XFAIL,
-        SKIPPED,
+        ERROR,      # red
+        FAILED,     # red
+        UNKNOWN,    # black
+        PASSED,     # green
+        UNSTABLE,   # yellow
+        None        # `status_override` is default to None
     )
 
-    STATUS_PROPAGATION = {
+    # And we map detailed status to primary status when we propagate upward
+    # or decide an entry should be considered passed/failed/unstable/unknown
+    STATUS_MAPPING = {
         ERROR: ERROR,
         FAILED: FAILED,
-        INCOMPLETE: INCOMPLETE,
+        INCOMPLETE: FAILED,
+        XPASS_STRICT: FAILED,
+        UNKNOWN: UNKNOWN,
         PASSED: PASSED,
-        SKIPPED: SKIPPED,
-        READY: READY,
-        RUNNING: RUNNING,
-        XFAIL: PASSED,
-        XPASS: PASSED,
-        None: None  # `status_override` can be None in merging report,
+        SKIPPED: UNSTABLE,
+        XFAIL: UNSTABLE,
+        XPASS: UNSTABLE,
+        UNSTABLE: UNSTABLE,
+        None: None  # `status_override` is default to None
```

Comment:
Again, we should not have a `None` status here. I think that the default status should be UNKNOWN (e.g. before tests are run, in interactive mode).

Reply:
status_override could be none, and when we merge 2 entries, we could be deciding the precedence between None and another status

Reply:
If status_override is None, the status comes from the child entries no? I don't think there should be any case where the status of a report node is returned as `None`.

Reply:
When we merge 2 entries (e.g multitest parts), we explicitly check status_override not the status property.